src/java.base/share/classes/java/lang/ref/Reference.java

```diff
@@ -325,14 +325,41 @@ public void runFinalization() {
      * been cleared, either by the program or by the garbage collector, then
      * this method returns {@code null}.
      *
+     * @apiNote
+     * This method returns a strong reference to the referent. This may cause
+     * the garbage collector to treat it as strongly reachable until some later
+     * collection cycle.  The {@link #refersTo(Object) refersTo} method can be
+     * used to avoid such strengthening when testing whether some object is
+     * the referent of a reference object; that is, use {@code ref.refersTo(obj)}
+     * rather than {@code ref.get() == obj}.
+     *
      * @return   The object to which this reference refers, or
      *           {@code null} if this reference object has been cleared
+     * @see refersTo
      */
     @IntrinsicCandidate
     public T get() {
         return this.referent;
     }
 
+    /**
+     * Tests if this reference object refers to {@code obj}.  If {@code obj} is
+     * {@code null}, this method returns {@code true} if this reference object
+     * refers to {@code null} or has been cleared.
+     *
+     * @param   obj the object to compare with this reference object's
+     *          referent, or {@code null}.
+     * @return  {@code true} if and only if {@code obj} is the referent
+     *          of this reference object
+     * @since 16
+     */
+    public final boolean refersTo(T obj) {
+        return refersTo0(obj);
+    }
+
+    /* Type-erased implementation of refersTo(). */
+    native boolean refersTo0(Object o);
```

Comment:
Is there any reason why it can't use the parameter type like `refersTo0(T o)`?  Same for `PhantomReference::refersTo0(T o)`

Reply:
But should `refersTo` takes a `T`? I mean - if I have a `Reference<?> ref`, the compiler will not let me query `ref.refersTo(o)`; is this what we want?

Reply:
@dfuch I would expect the users of `refersTo` should know the type of the referent to be compared, i.e. it should know the parameter type of a Reference object.   Do you have any specific use case in mind that you have a `Reference<?> ref` but wants to compare a referent of unknown type?

Reply:
@mlchung  I have often used a `Reference<?>` in tests - but my main usage there would be to call `ref.refersTo(null)` which works in all cases.
My main concern here however is that using `T` in `refersTo` seems to go against the advertised usage of the method - I mean - if I have an object `obj` of type unknown, I can always do `obj == ref.get()` whatever the parameter type of `ref` is. But I won't be able to call `ref.refersTo(obj)` unless I use raw types and suppress warnings. So I just wanted to check that this was intentional.

Reply:
For the common cases, the application should know the type of the referent and using `T` in `refersTo` will benefit from the compiler type checking.   For the unknown type case, cast to `Reference<Object>` is not ideal but reasonable?  something like this:

```
Reference<Object> r = (Reference<Object>) ref;
r.refersTo(obj);
```

Reply:
That sounds reasonable to me. Thanks for looking into it.

Reply:
I just want to note that if you have a `Reference<SomeType> ref` at hand, you can not just do:
```
Referemce<Object> r = (Reference<Object>) ref;
```
...since those generic types are not related. You have to do something like:

```
@SuppressWarnings({"unchecked", "rawtypes"})
Referemce<Object> r = (Reference) ref;
```
which is very unfortunate. Comparing this method with for example `Collection.contains(Object element)`, you can see that Collection API has made a decision to not bother with T here. That was also due to keeping old code compatible when migrating from pre-generics Java to generified Collection, but as @dfuch noted, we have a migration story here too. We will be migrating `obj == ref.get()` to `ref.refersTo(obj)` ... Mind you that this is a boolean expression fragment which might be written inline surrounded with other parts of expression. So you'll be forced to split that into assignment with @SuppressWarnings and an expression or you will have to force the whole expression or method to @SuppressWarnings. I don't know if type "safety" is forth it here.

Reply:
Can you point to some existing code that wants to call `refersTo` to compare a `Reference` of raw type with an object of unknown type?   Reference instances should not be leaked and so I don't see very common that caller of `Reference::get` does not know the referent's type.   It also depends on the `refersTo` check against `null` vs an object.  Any known use case would be helpful.

FWIW, when converting a few use of `Reference::get` to `refersTo` in JDK, there is only one case (`equals(Object o)` method that needs the cast.

http://cr.openjdk.java.net/~mchung/jdk15/webrevs/8188055/jdk-use-refersTo/index.html

Reply:
Is there a consensus on this issue?  It's not clear whether Daniel and Peter have agreed with Mandy's responses or have just not yet responded with further discussion.

Reply:
@mlchung I don't have many known use cases, but how about WeakHashMap.containsKey(Object key) for example? Currently `WeakHashMap.Entry<K, V> extends  WeakReference<Object>` but it would be more type safe if it extended `WeakReference<K>`. In that case an `entry.refersTo(key)` would not work...

Reply:
Some thoughts regarding the parameter type of refersTo. Summary: I think `refersTo(T)` is fine and that we don't want to change it to `refersTo(Object)`.

I don't think we have a migration issue similar to generifying collections, where there was a possibility of changing `contains(Object)` to `contains(E)`. If that had been done, it would have been a source compatibility issue, because changing the signature of the method potentially affects existing code that calls the method. That doesn't apply here because we're adding a new method.

The question now falls to whether it's preferable to have more convenience with `refersTo(Object)` or more type-safety with `refersTo(T)`. With the generic collections issue, the migration issue probably drove the decision to keep `contains(Object)`, but this has resulted in a continual set of complaints about the lack of an error when code passes an instance of the "wrong" type. I think that kind of error is likely to occur with `refersTo`. Since we don't have a source compatibility issue here, we can choose the safer API and avoid this kind of problem entirely.

The safer API does raise the possibility of having to add inconvenient unchecked casts and local variables in certain places, but I think Mandy's comment about the code already having a reference of the "right" type is correct. Her prototype webrev linked above shows that having to add unchecked casts is fairly infrequent.

Reply:
> Some thoughts regarding the parameter type of refersTo. Summary: I think `refersTo(T)` is fine and that we don't want to change it to `refersTo(Object)`.
> 
I agree that we don't have a migration problem here that collections had. So let it be `refersTo(T)` then.

Reply:
I agree as well.